### PR TITLE
Feature/em tracking

### PR DESCRIPTION
**Pull request by giadarol**

### CODE REVIEW

File: buildup_simulation.py [hunk 1]
```diff
@@ -61,7 +61,7 @@
 class BuildupSimulation(object):
 
     def __init__(self, pyecl_input_folder='./', skip_beam=False, skip_spacech_ele=False,
-                 skip_pyeclsaver=False, ignore_kwargs=[], spacech_ele=None, **kwargs):
+                 skip_pyeclsaver=False, lorentz_boost = True, ignore_kwargs=[], spacech_ele=None, **kwargs):
```
Comment (giadarol) [inline buildup_simulation.py]: @lgiacome This should be handled in the input file

File: buildup_simulation.py [hunk 2]
```diff
@@ -85,6 +85,9 @@ def __init__(self, pyecl_input_folder='./', skip_beam=False, skip_spacech_ele=Fa
         self.chamb = cloud_list[0].impact_man.chamb
         self.checkpoint_folder = checkpoint_folder
 
+        #!#!
+        self.lorentz_boost = lorentz_boost
```
Comment (giadarol) [inline buildup_simulation.py]: @lgiacome should we change the name?

File: buildup_simulation.py [hunk 3]
```diff
@@ -165,8 +168,16 @@ def sim_time_step(self, beamtim_obj=None, Dt_substep_custom=None, N_sub_steps_cu
                     Ex_n_beam += Ex_n_secbeam
                     Ey_n_beam += Ey_n_secbeam
 
-            ## Compute electron space charge electric field
-            Ex_sc_n, Ey_sc_n = spacech_ele.get_sc_eletric_field(MP_e)
+            #!#! Either compute electromagnetic field or electrostatic
+            if self.lorentz_boost:
+                Ex_sc_n, Ey_sc_n, Bx_sc_n, By_sc_n, Bz_sc_n = spacech_ele.get_sc_b_field(MP_e)
```
Comment (giadarol) [inline buildup_simulation.py]: @lgiacome get_sc_b_field should become get_em_fields

File: buildup_simulation.py [hunk 4]
```diff
@@ -185,22 +196,26 @@ def sim_time_step(self, beamtim_obj=None, Dt_substep_custom=None, N_sub_steps_cu
...
             elif Dt_substep_custom is None and N_sub_steps_custom is None and not(beamtim.flag_unif_Dt):
                 # Dt from non-uniform beam profile
                 if self.config_dict['track_method'] not in ['Boris', 'BorisMultipole']:
                     raise ValueError("""track_method should be 'Boris' or 'BorisMultipole' to use custom substeps!""")
                 Dt_substep_target = cloud.dynamics.Dt / cloud.dynamics.N_sub_steps
                 N_substeps_curr = np.round(beamtim.Dt_curr / Dt_substep_target)
                 Dt_substep_curr = beamtim.Dt_curr / N_substeps_curr
-                MP_e = dynamics.stepcustomDt(MP_e, Ex_n, Ey_n, Dt_substep=Dt_substep_curr, N_sub_steps=N_substeps_curr)
+                Ez_n = 0
+                MP_e = dynamics.stepcustomDt(MP_e, Ex_n, Ey_n, Ez_n, Bx_sc_n, By_sc_n, Bz_sc_n, Dt_substep=Dt_substep_curr, N_sub_steps=N_substeps_curr)
```
Comment (giadarol) [inline buildup_simulation.py]: @lgiacome Remember to test both branches of the if

File: buildup_simulation.py [hunk 5]
```diff
@@ -185,22 +196,26 @@ def sim_time_step(self, beamtim_obj=None, Dt_substep_custom=None, N_sub_steps_cu
...
             else:
                 # Custom steps and substeps provided as arguments of sim_time_step
                 if self.config_dict['track_method'] not in ['Boris', 'BorisMultipole']:
                     raise ValueError("""track_method should be 'Boris' or 'BorisMultipole' to use custom substeps!""")
-                MP_e = dynamics.stepcustomDt(MP_e, Ex_n, Ey_n, Dt_substep=Dt_substep_custom, N_sub_steps=N_sub_steps_custom)
+                Ez_n = 0
+                MP_e = dynamics.stepcustomDt(MP_e, Ex_n, Ey_n, Ez_n, Bx_sc_n, By_sc_n, Bz_sc_n, Dt_substep=Dt_substep_custom, N_sub_steps=N_sub_steps_custom)
```
Comment (giadarol) [inline buildup_simulation.py]: @lgiacome Remember to test all branches of the if

File: dynamics_Boris_multipole.py [hunk 6]
```diff
@@ -83,7 +83,12 @@ def __init__(self, Dt, N_sub_steps=1, B_multip=None, B_skew=None):
         print "N_subst_init=%d" % self.N_sub_steps
 
     #@profile
-    def step(self, MP_e, Ex_n, Ey_n, Ez_n=0.):
+    def step(self, MP_e, Ex_n, Ey_n, Ez_n=0., Bx_n=[0.], By_n=[0.], Bz_n=[0.]):
+
+        if (Bx_n == [0.]).all() and (By_n == [0.]).all() and (Bz_n == [0.]).all():
+            custom_B = 0
+        else:
+            custom_B = 1
```
Comment (giadarol) [inline dynamics_Boris_multipole.py]: @lgiacome Defaults should be arrays, not lists

File: init.py [hunk 7]
```diff
@@ -76,7 +76,7 @@
 import dynamics_Boris_multipole as dynmul
 
 import MP_system as MPs
-import space_charge_class as scc
+import space_charge_class_electromagnetic as scc
```
Comment (giadarol) [inline init.py]: Handle properly from input files

File: space_charge_class_electromagnetic.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+from space_charge_class import space_charge
+from scipy.constants import c, epsilon_0, mu_0
+
+na = lambda x: np.array([x])
+
+
+class space_charge_electromagnetic(space_charge, object):
+
+    def __init__(self, chamb, Dh, Dt_sc=None, PyPICmode='FiniteDifferences_ShortleyWeller' , sparse_solver='scipy_slu',
+                 f_telescope=None, target_grid=None, N_nodes_discard=None, N_min_Dh_main=None, gamma = 479.):
```
Comment (giadarol) [inline space_charge_class_electromagnetic.py]: @lgiacome change to gamma = None; never use magic numbers

File: space_charge_class_electromagnetic.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+                 epsilon_0 * mu_0 * MP_e.nel_mp[0:MP_e.N_mp] * MP_e.vy_mp[0:MP_e.N_mp],
+                charge=MP_e.charge, flag_add=not(flag_reset))
+
+        # solve
+        if flag_solve:
+            self.PyPICobj.solve()
+            self.PyPICobj.solve_states([self.state_Ax, self.state_Ay])
+
+
+    def get_sc_b_field(self, MP_e):
```
Comment (giadarol) [inline space_charge_class_electromagnetic.py]: @lgiacome rename

File: space_charge_class_electromagnetic.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+        _, dAx_dy = self.state_Ax.gather(MP_e.x_mp[0:MP_e.N_mp], MP_e.y_mp[0:MP_e.N_mp])
+        dAy_dx, _ = self.state_Ay.gather(MP_e.x_mp[0:MP_e.N_mp], MP_e.y_mp[0:MP_e.N_mp])
+        dphi_dx, dphi_dy = self.PyPICobj.gather(MP_e.x_mp[0:MP_e.N_mp], MP_e.y_mp[0:MP_e.N_mp])
+
+        Ex_prime = self.gamma*dphi_dx
+        Ey_prime = self.gamma*dphi_dy
+
+        dAz_dx = -self.gamma*self.beta*c*epsilon_0*mu_0*dphi_dx
```
Comment (giadarol) [inline space_charge_class_electromagnetic.py]: @lgiacome Let's try to be consistent with the notation of the note: distinguishing phi from Phi

**Comment by lgiacome** [comment]
> I think that this is wrong :-)

Me too

File: pyecloud_saver.py [hunk 11]
```diff
@@ -991,33 +1011,91 @@ def _sc_video_init(self, flag_sc_movie):
...
+        self.Fefy_video = None
+        self.Fbfx_video = None
+        self.Fbfy_video = None
+        self.Fbfz_video = None
         self.t_efield_video = None
 
-    def _sc_video_save(self, spacech_ele, beamtim):
-        #save efield video
+    def _sc_video_save(self, spacech_ele, beamtim, MP_e):
         if self.flag_sc_video:
```
Comment (giadarol) [inline pyecloud_saver.py]: @lgiacome Let's remove this part, it will be painful to keep it alive

**Comment by giadarol** [comment]
Will be handled in #113

**Pull request closed without merging**